db/migrations/20190711091453_create_flip_table.sql

```diff
@@ -135,256 +166,657 @@ COMMENT ON FUNCTION flip_bid_time_created
 
 
 -- +goose StatementBegin
-CREATE OR REPLACE FUNCTION maker.insert_updated_flip_guy() RETURNS TRIGGER AS
+CREATE OR REPLACE FUNCTION maker.insert_new_flip_guy(new_diff maker.flip_bid_guy) RETURNS VOID AS
+$$
+WITH diff_block AS (
+    SELECT block_number, block_timestamp
+    FROM public.headers
+    WHERE id = new_diff.header_id
+)
+INSERT
+INTO maker.flip (bid_id, address_id, block_number, guy, tic, "end", lot, bid, usr, gal, tab, updated, created)
+VALUES (new_diff.bid_id,
+        new_diff.address_id,
+        (SELECT block_number FROM diff_block),
+        new_diff.guy,
+        flip_bid_tic_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_end_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_lot_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_bid_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_usr_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_gal_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_tab_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        (SELECT api.epoch_to_datetime(block_timestamp) FROM diff_block),
+        flip_bid_time_created(new_diff.address_id, new_diff.bid_id))
+ON CONFLICT (block_number, bid_id, address_id) DO UPDATE SET guy = new_diff.guy;
 $$
-BEGIN
-    WITH diff_block AS (
-        SELECT block_number, block_timestamp
+    LANGUAGE sql;
+-- +goose StatementEnd
+
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.update_flip_guys_until_next_diff(new_diff maker.flip_bid_guy) RETURNS VOID
```

Comment:
The flip table is basically a collection of snapshots, each representing the state of a bid at a given block. So when a diff from the past is inserted, we need to update all the snapshots which would've been affected by the diff that was inserted. This trigger's job is to figure out which snapshots would be affected, and update them accordingly.

Using `flip_bid_guy` as an example: the only snapshots we care about updating are the ones from blocks in between the `flip_bid_guy` from the past (which was just inserted), and the next-most-recent `flip_bid_guy` diff. That's because the next one will essentially overwrite the `guy` field for that bid, but we still want a record of what the bid's state was before that later diff happened.

Hope that makes sense! Lmk if I can clarify anything.

Reply:
🙏 thank you, that makes sense.